F/173809 unified search

drspacemanphd opened the pull request:
[173809](https://esriarlington.tpondemand.com/entity/173809-spike-prototype-for-unified-search-interface).

This PR offers an initial implementation of a Unified Search interface. Specific context and more information can be found in the [package README](https://github.com/Esri/hub.js/blob/f/173809-unified-search/packages/search/README.md).

I've deployed four sandbox UIs in which the Content Library page has been modified to exclusively use the proposed interface to fetch content. 

1. The `mab-search-portal` sandbox uses the `searchContent` function configured with `isPortal: true`
2. The `mab-search-hub` sandbox uses the `searchContent` function configured with `isPortal: false`
3. The `mab-search-portal-service` sandbox uses the `ContentSearchService` configured with `isPortal: true`
4. The `mab-search-hub-service` sandbox uses the `ContentSearchService` configured with `isPortal: false`

https://reduce-noise-pollution-on-mass-ave-qa-pre-hub.hubqa.arcgis.com/content/edit?uiVersion=mab-search-portal
https://qa-pre-hub.hubqa.arcgis.com/content/edit?uiVersion=mab-search-portal
https://reduce-noise-pollution-on-mass-ave-qa-pre-hub.hubqa.arcgis.com/content/edit?uiVersion=mab-search-hub
https://qa-pre-hub.hubqa.arcgis.com/content/edit?uiVersion=mab-search-hub
https://reduce-noise-pollution-on-mass-ave-qa-pre-hub.hubqa.arcgis.com/content/edit?uiVersion=mab-search-portal-service
https://qa-pre-hub.hubqa.arcgis.com/content/edit?uiVersion=mab-search-portal-service
https://reduce-noise-pollution-on-mass-ave-qa-pre-hub.hubqa.arcgis.com/content/edit?uiVersion=mab-search-hub-service
https://qa-pre-hub.hubqa.arcgis.com/content/edit?uiVersion=mab-search-hub-service

trescube reviewed (approved):
This is a great initial implementation!  I'm looking forward to see how it progresses.

drspacemanphd commented on packages/search/src/content/helpers/convert-request-to-portal-params.ts:
```diff
@@ -0,0 +1,183 @@
+
+function processFilter(request: IContentSearchRequest): string {
+  const filter: IContentSearchFilter = request.filter || {};
+  const filters: string[] = Object.keys(filter).reduce(
```
`Object.entries` I believe is ES8 and the target for Hub.js is ES6

trescube replied:
Ah, cool, figured there was a reason.  Thx!

tomwayson commented on packages/search/README.md:
````diff
@@ -55,6 +55,113 @@ const { agoSearch } = require('@esri/hub-search')
 ```
 
+### Content Search
+The `ContentSearchService` provides an API to search for content from the Portal API only or from both ArcGIS Online and Hub Indexer's V3 Search API (hereafter known as `V3 Search API`), depending on how the service is configured.
````
I think how you instantiate this class in an Ember CP is really clever. 

https://github.com/ArcGIS/opendata-ui/compare/content-search-prototype#diff-3540ef806c42f464ef6f89f0d93a839089a876217c63cf6f277632ef68701f73R235-R237

That helps alleviate the concerns that I raised in https://github.com/Esri/hub.js/pull/445#discussion_r568821838

Now the CP above is on a route, and very likely we'd want to be able to search from different routes so we'd want an Ember service. I could see how we could have a `hub` service that we could inject anywhere (i.e. `@service hub`) and that in turn could then expose instances of these Hub.js "service" classes to the app as CPs like:

```js
// services/hub.js

@service appSettings;
@service session;

get contentSearchService () {
  // NOTE: does using class syntax cache this instance the way using CPs on an Ember.Object would?
  return new ContentSearchService({ portalSharingUrl: this.appSettings.sharingUrl, isPortal: this.appSettings.isPortal, userSession: this.session.authMgr });
}
```

then in some route:

```js
// some-route.js

@service hub;

searchContent (filter) {
  return this.hub.contentSearchService.search({ filter });
} 
```

Hm... after I write all that out it still feels a little cumbersome.

At the end of the day what we want is `this.hub.content.search({ filters, options})` and we're still not quite there yet, but I think that has more to do w/ the structure of Hub.js packages than what you are doing in this PR.

So, all that being said, introducing a `class` _at this time_ requires us to go through some otherwise unnecessary gymnastics. I'd much prefer if this PR just exposed the stateless fns that could later be composed into a `HubContent()` class that would then expose `.search()`, `.fetch()`, etc.

tomwayson replied:
@dbouwman  and @mjuniper what do you guys think about this?

drspacemanphd replied:
> Minor naming convention preference. I'm not really fond of use the `Service` suffix for classes like this and the `UserService` proposed in #445. The only reason is b/c if/when we used them in the Ember app it's going to cause confusion whether or not we are talking about an Ember service or an instance of the Hub.js class.

I'm not sure I understand this given the Intellisense IDEs have today and the prevalence of named imports.

tomwayson reviewed (commented):
Impressive work @drspacemanphd!

I think your system for building either Hub or Portal API params from a common `filter` object is well done. I was hoping we'd be able to leverage  rest-js' [SearchQueryBuilder](https://esri.github.io/arcgis-rest-js/api/portal/SearchQueryBuilder/) for this, but I after our offline discussion, I understand why you didn't.

I still think that we're out over our skis a bit by introducing classes at this time. I definitely do see the value of classes, but until we are able to expose the API we want (`hub.content.search()` and `hub.content.fetch()`) I don't see any value in adding them now. If this PR only introduced the stateless fns I'd approve it.

drspacemanphd commented:
> Impressive work @drspacemanphd!
> 
> I think your system for building either Hub or Portal API params from a common `filter` object is well done. I was hoping we'd be able to leverage rest-js' [SearchQueryBuilder](https://esri.github.io/arcgis-rest-js/api/portal/SearchQueryBuilder/) for this, but I after our offline discussion, I understand why you didn't.
> 
> I still think that we're out over our skis a bit by introducing classes at this time. I definitely do see the value of classes, but until we are able to expose the API we want (`hub.content.search()` and `hub.content.fetch()`) I don't see any value in adding them now. If this PR only introduced the stateless fns I'd approve it.

@tomwayson I have update the PR to keep the `ContentSearchService` but to also include a `searchContent` standalone function. I also refactored the enterprise and hub content search methods by extracting the core logic into non-exported standalone functions. This way the actual search logic is only in one place. The client can now simply call `await searchContent({ filters, options })`.

drspacemanphd commented on packages/search/src/types/search-service.ts:
```diff
@@ -0,0 +1,11 @@
+import { UserSession } from "@esri/arcgis-rest-auth";
+
+export interface ISearchServiceParams {
```
I'd rather limit the scope of options that users can (or have to) provide, but I have taken your advice to rename these properties for consistency.

drspacemanphd commented on packages/search/src/types/content.ts:
```diff
@@ -0,0 +1,66 @@
+
+export interface IContentSearchOptions {
+  session?: UserSession;
+  portalSharingUrl?: string;
```
See above. I have renamed.

drspacemanphd commented on packages/search/src/content/helpers/convert-hub-response.ts:
```diff
@@ -0,0 +1,144 @@
+  serviceSession?: UserSession
+): () => Promise<IContentSearchResponse> {
+  return (userSession?: UserSession) => {
+    const authentication: UserSession = userSession || serviceSession;
```
This is actually the case, but it wasn't very clear. I have updated the PR to hopefully make this a bit more intuitive. I've also updated the README.

dbouwman commented on packages/search/README.md:
```diff
@@ -55,6 +55,117 @@ const { agoSearch } = require('@esri/hub-search')
+  - `hasNext`: a boolean indicating if there are more pages of results
+  - `next`: a function that, when invoked, will fetch the next page of results
+
+Importantly, one can optionally provide a UserSession instance to the `next` function invocation to be used for authenticated searches
```
Not quite following this... what scenario would you issue a search w/o a `session`, and then need/want to add it on the `.next` invocation? I assume `.next` is a closure over some state, and if a `session` was provided on the initial call we could hold that in the closure? Of course this gives me the memory-leak-willies, b/c we're holding class instances in closures of indeterminate lifetimes... but @mikeringrose  tells me this is "ok", so I'm trying to get over that ;)

drspacemanphd replied:
I originally included it in an effort to keep it as "stateless" as possible, ie. you can always pass in authentication with each invocation if you want to.

I'm not an expert in Ember and its memory management but given what is currently in the UI, I don't believe we would be at significant risk. It looks like old `searchResults`, which would include the `next` function, should be dereferenced, but happy to chat more about it.

```
async fetchMore () {
  const currentResults = this.model.searchResults.results;
  const res = await this.model.searchResults.next();
  res.results = [...currentResults, ...res.results];
  this.set('model.searchResults', res);
},
```

dbouwman commented on packages/search/README.md:
```diff
@@ -55,6 +55,117 @@ const { agoSearch } = require('@esri/hub-search')
+**Importantly, this means that, when searching the V3 Search API, supplying filter values for these fields will be interpeted as an exact search, leading to very few results. Fuzzy searching should be done via the search term**
+
+3. NOTs cannot be combined with non-NOT operators on the same field
+`(tag: TAG_ONE AND NOT tag: TAG_TWO)`
```
Is there a way to accomplish this though? like `(tag: TAG_ONE AND -tag: TAG_TWO)` Not that we do this much with tags, but we regularly do snarly stuff like this with `typekeywords` - granted that's usually in an item picker, vs doing things with "content" per-se, but I know @ajturner has a expressed vision (at least in the past) where even item pickers are powered by the "Hub Search".

drspacemanphd replied:
Unfortunately not with the current API. The request syntax is very different from Portal and doesn't lend to this or the other types of searches mentioned in the caveats section. In researching search across the application, I mentioned this to @thomas-hervey, particularly regarding the `typekeywords` you mention. A really good example of searches that the V3 API cannot current support are found in the "Add Existing Content" modal.

dbouwman replied:
Ok - good to know that until this is addressed, Item Pickers should likely be powered by the Portal API directly.

drspacemanphd replied:
I have to agree, although I have to discuss with @thomas-hervey a bit more, particularly as we begin to add other sources such as Event Feature Layers to the search

dbouwman reviewed (changes requested):
I added a few inline questions re: the search syntax, but generally this seems pretty slick.

From the ember app perspective, I'm assuming we'd create a `ContentService` in ember, and inside of that, lean on the `searchContent` function. Since Ember services are singletons that are created very early in the app boot cycle (well before we can process any existing session), leaning on the function vs the class side-steps issues of syncing the session into the instance of the hub.js `ContentSearchService` that would reside in the service. While we could use this directly in a route/controller or component, stubbing external functions brought in via ember-auto-import is snarly at best... and generally involves resorting to fetchMock... which is horribles. Ember services are much simpler.

Last thing and the actual "ask" - since doc is auto-generated, all exported functions/classes/interfaces should have jsdoc/tsdoc style comments. I'd also recommend adding a guide to the `/docs/src/guides` folder... I _believe_ that is enough to have it appear in the doc site, but we may need to edit some other file(s) as well. Let me know if you need help with that.

dbouwman commented:
Not sure why, but when I loaded this initially, I did not see the other conversations w/ Tom... or maybe I followed a link directly into the Files Changed tab... but I think we end up in the same place... this sort of functionality should be exposed into the ember app via an ember service. tis the ember-way.

dbouwman commented:
Also @drspacemanphd - if you don't have it, get the "Document This" vscode addin if you don't have another one - it picks up the typings etc and add it into the doc comment https://marketplace.visualstudio.com/items?itemName=oouo-diogo-perdigao.docthis

drspacemanphd commented:
> Not sure why, but when I loaded this initially, I did not see the other conversations w/ Tom... or maybe I followed a link directly into the Files Changed tab... but I think we end up in the same place... this sort of functionality should be exposed into the ember app via an ember service. tis the ember-way.

I mentioned to Tom offline, but I mostly used a CP as a quick UI functionality test but noticed that it was only being instantiated once. It can be injected as a service.

drspacemanphd commented:
> I added a few inline questions re: the search syntax, but generally this seems pretty slick.
> 
> From the ember app perspective, I'm assuming we'd create a `ContentService` in ember, and inside of that, lean on the `searchContent` function. Since Ember services are singletons that are created very early in the app boot cycle (well before we can process any existing session), leaning on the function vs the class side-steps issues of syncing the session into the instance of the hub.js `ContentSearchService` that would reside in the service. While we could use this directly in a route/controller or component, stubbing external functions brought in via ember-auto-import is snarly at best... and generally involves resorting to fetchMock... which is horribles. Ember services are much simpler.
> 
> Last thing and the actual "ask" - since doc is auto-generated, all exported functions/classes/interfaces should have jsdoc/tsdoc style comments. I'd also recommend adding a guide to the `/docs/src/guides` folder... I _believe_ that is enough to have it appear in the doc site, but we may need to edit some other file(s) as well. Let me know if you need help with that.

Thanks for going through this and for the suggestions. I've added what I believe is satisfactory documentation to exported functions/classes/interfaces and copied most of what's in the README into a guide as you suggested.

dbouwman reviewed (approved):
Thanks @drspacemanphd - this is a big step forward - much appreciated and thanks for your patience w/ the review process!

There are still a few exported functions w/o comments (`processPage`, `decodePage` are ones I saw) - pls add and then feel free to merge and release.

drspacemanphd commented:
> Thanks @drspacemanphd - this is a big step forward - much appreciated and thanks for your patience w/ the review process!
> 
> There are still a few exported functions w/o comments (`processPage`, `decodePage` are ones I saw) - pls add and then feel free to merge and release.

Whoops, got them this time.

tomwayson reviewed (approved):
Thanks for making the requested changes @drspacemanphd, this LGTM!
